src/ExceptionTypes.f90

```diff
@@ -0,0 +1,370 @@
+      !> @copydetails ExceptionTypes::setVerboseMode_ExceptionTypeBase
+      PROCEDURE,PASS :: setVerboseMode => setVerboseMode_ExceptionTypeBase
+      !> @copybrief ExceptionTypes::resetCounter_ExceptionTypeBase
+      !> @copydetails ExceptionTypes::resetCounter_ExceptionTypeBase
+      PROCEDURE,PASS :: resetCounter => resetCounter_ExceptionTypeBase
+      !> @copybrief ExceptionTypes::setCounter_ExceptionTypeBase
+      !> @copydetails ExceptionTypes::setCounter_ExceptionTypeBase
+      PROCEDURE,PASS :: setCounter => setCounter_ExceptionTypeBase
+      !> @copybrief ExceptionTypes::getCounter_ExceptionTypeBase
+      !> @copydetails ExceptionTypes::getCounter_ExceptionTypeBase
+      PROCEDURE,PASS :: getCounter => getCounter_ExceptionTypeBase
+      !> @copybrief ExceptionTypes::logMessage_ExceptionTypeBase
+      !> @copydetails ExceptionTypes::logMessage_ExceptionTypeBase
+      PROCEDURE,PASS :: logMessage => logMessage_ExceptionTypeBase
+      !> @copybrief ExceptionTypes::getTag_ExceptionTypeBase
+      !> @copydetails ExceptionTypes::getTag_ExceptionTypeBase
+      PROCEDURE,PASS :: getTag => getTag_ExceptionTypeBase
+      !>
+      !> abstract interface functions (absintfc)
+      !>
+      !> @copybrief ExceptionTypes::et_genprefix_absintfc
+      !> @copydetails ExceptionTypes::et_genprefix_absintfc
+      PROCEDURE(et_genprefix_absintfc),DEFERRED,PASS :: genPrefix
+
+      !> @
+  ENDTYPE ExceptionTypeBase
+
+!> Concrete type for Information exception
+  TYPE,EXTENDS(ExceptionTypeBase) :: ExceptionTypeInformation
+    CONTAINS
+      PROCEDURE,PASS :: genPrefix => genPrefix_ExceptionTypeInformation
+  ENDTYPE ExceptionTypeInformation
+
+!> Concrete type for Warning exception
+  TYPE,EXTENDS(ExceptionTypeBase) :: ExceptionTypeWarning
+    CONTAINS
+      PROCEDURE,PASS :: genPrefix => genPrefix_ExceptionTypeWarning
+  ENDTYPE ExceptionTypeWarning
+
+!> Concrete type for Debug exception
+  TYPE,EXTENDS(ExceptionTypeBase) :: ExceptionTypeDebug
+    CONTAINS
+      PROCEDURE,PASS :: genPrefix => genPrefix_ExceptionTypeDebug
+  ENDTYPE ExceptionTypeDebug
+
+!> Concrete type for Error exception
+  TYPE,EXTENDS(ExceptionTypeBase) :: ExceptionTypeError
+    CONTAINS
+      PROCEDURE,PASS :: genPrefix => genPrefix_ExceptionTypeError
+  ENDTYPE ExceptionTypeError
+
+!> Concrete type for Fatal exception
+  TYPE,EXTENDS(ExceptionTypeBase) :: ExceptionTypeFatal
+    CONTAINS
+      PROCEDURE,PASS :: genPrefix => genPrefix_ExceptionTypeFatal
+  ENDTYPE ExceptionTypeFatal
+!
+ABSTRACT INTERFACE
+  !> @brief Abstract interface for appending prefix to error msg
+  SUBROUTINE et_genprefix_absintfc(this,prefix)
+    IMPORT :: ExceptionTypeBase
+    IMPORT :: EXCEPTION_MAX_MESG_LEN
+    CLASS(ExceptionTypeBase),INTENT(INOUT) :: this
+    CHARACTER(LEN=EXCEPTION_MAX_MESG_LEN),INTENT(INOUT) :: prefix
+  ENDSUBROUTINE et_genprefix_absintfc
+ENDINTERFACE
+!
+!===============================================================================
+  CONTAINS
```

Comment:
Everything after this `CONTAINS` statement needs to be unindented an additional 2 spaces on top of the 2 I mentioned above, so that `SUBROUTINE`, `FUNCTION`, and their matching `END` statements start in the first column.